Add new table `system.codecs`

Ergus opened the pull request:
Closes #81525

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a new system table `system.codecs` to introspect the available codecs. (issue #81525)

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

Ergus committed:
Introspect the list of codecs and their properties.

* src/Storages/System/StorageSystemCodecs.{h,cpp} : New files with the new table class.
* src/Storages/System/StorageSystemCodecs.{h,cpp} (getMutableColumns) : New function to get columns information accessing the private members in CompressionCodecFactory
* src/Storages/System/attachSystemTables.cpp : Attach the new table to system.

Ergus committed:
Add documentation column "notes" to the system.codecs table

Add a new virtual function getNotes that returns documentation for the codec taken from:
https://kb.altinity.com/altinity-kb-schema-design/codecs/

* src/Compression/ICompressionCodec.h (getData) : New interface function. And update in all derived classes.

Ergus committed:
Add system.codecs table test

* tests/queries/0_stateless/01222_system_codecs.{sql, reference} : New test files

Ergus committed:
Add documentation for site and update CHANGELOG.

(docs/en/operations/system-tables/codecs.md) : New file
(CHANGELOG.md) : Add a line introducing the new feature.

novikd commented on src/Compression/CompressionFactory.h:
```diff
@@ -74,6 +75,8 @@ class CompressionCodecFactory final : private boost::noncopyable
     /// Get codec by name with optional params. Example: LZ4, ZSTD(3)
     CompressionCodecPtr get(const String & compression_codec) const;
 
+    void getMutableColumns(MutableColumns & res_columns) const;
```
Let's rename it and add a comment with a description. I'd recommend something like `fillCodecDescriptions`.

Ergus replied:
Done

Ergus commented on docs/en/operations/system-tables/codecs.md:
```diff
@@ -0,0 +1,47 @@
+The `system.codecs` table contains the following columns (the column type is shown in brackets):
+
+- `codec` ([String](../../sql-reference/data-types/string.md)) — Codec name.
+- `method_byte` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — Byte which indicates codec in compressed file.
```
Agree, I will remove it then

Ergus commented on docs/en/operations/system-tables/codecs.md:
```diff
@@ -0,0 +1,47 @@
+- `floating_point_timeseries`([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is for floating point timeseries codec.
+- `delta_compression` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is for delta compression.
+- `experimental` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is experimental.
+- `notes` ([String](../../sql-reference/data-types/string.md)) — Performance notes.
```
Yes, I added the docs column in a separated commit because of that. But the @alexey-milovidov request explicitly says: "and provide a built-in documentation"

Blargian replied:
@rschu1ze I am for generating as much reference documentation as possible from system tables.

rschu1ze replied:
Okay, sounds good, let's keep it then :-)

Ergus commented on src/Compression/CompressionFactory.cpp:
```diff
@@ -112,6 +115,28 @@ CompressionCodecPtr CompressionCodecFactory::get(uint8_t byte_code) const
     return family_code_and_creator->second({}, nullptr);
 }
 
+void CompressionCodecFactory::getMutableColumns(MutableColumns & res_columns) const
```
We could but this follows the prototype of other similar functions for the same purpose.

 The `MutableColumns` object received by reference seems to be pre-initialized with the information from `getColumnsDescription`. I don't like this design because it makes the 3 functions inter-dependent; but I didn't want to break the current pattern followed by many other system tables.

Ergus commented on docs/en/operations/system-tables/codecs.md:
```diff
@@ -0,0 +1,47 @@
+
+The `system.codecs` table contains the following columns (the column type is shown in brackets):
+
+- `codec` ([String](../../sql-reference/data-types/string.md)) — Codec name.
```
Sure

Ergus commented on docs/en/operations/system-tables/codecs.md:
```diff
@@ -0,0 +1,47 @@
+- `compression` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is a compression.
+- `generic_compression` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is a generic compression.
+- `encription` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec encrypts.
+- `floating_point_timeseries`([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is for floating point timeseries codec.
```
Yes, not sure why the virtual function had such a complex name...

Ergus commented on src/Compression/CompressionFactory.cpp:
```diff
@@ -112,6 +115,28 @@ CompressionCodecPtr CompressionCodecFactory::get(uint8_t byte_code) const
     return family_code_and_creator->second({}, nullptr);
 }
 
+void CompressionCodecFactory::getMutableColumns(MutableColumns & res_columns) const
```
I renamed it to `fillCodecDescriptions` is it good enough?

rschu1ze replied:
++

Ergus commented on tests/queries/0_stateless/01222_system_codecs.sql:
```diff
@@ -0,0 +1,5 @@
+-- Tags: no-object-storage
+select * from system.codecs where codec = 'LZ4';
```
I didn't do that because I was not sure if adding a new codec may "break" this test requiring updating the reference. The developer could check his codec's and won't double check of the others changed for some reason.
But Ok.

Ergus commented on CHANGELOG.md:
```diff
@@ -26,6 +26,7 @@
 * If reading from an `URL` involves multiple redirects, setting `enable_url_encoding` is correctly applied across all redirects in the chain. [#79563](https://github.com/ClickHouse/ClickHouse/pull/79563) ([Shankar Iyer](https://github.com/shankar-iyer)). Setting `enble_url_encoding` default value is now set to `false`. [#80088](https://github.com/ClickHouse/ClickHouse/pull/80088) ([Shankar Iyer]( ...
 
 #### New Feature
+* New system table `system.codecs` to introspect the list of codecs, their properties, and provide a built-in documentation. Closes [#81525](https://github.com/ClickHouse/ClickHouse/issues/81525) ([Jimmy Aguilar Mena](https://github.com/Ergus)).
```
Removed

Ergus commented on docs/en/operations/system-tables/codecs.md:
```diff
@@ -0,0 +1,47 @@
+
+- `codec` ([String](../../sql-reference/data-types/string.md)) — Codec name.
+- `method_byte` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — Byte which indicates codec in compressed file.
+- `compression` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is a compression.
```
Done

Ergus committed:
CR suggestions from @rschu1ze and @novikd

Ergus commented on docs/en/operations/system-tables/codecs.md:
```diff
@@ -0,0 +1,47 @@
+- `method_byte` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — Byte which indicates codec in compressed file.
+- `compression` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is a compression.
+- `generic_compression` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is a generic compression.
+- `encription` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec encrypts.
```
Hi:

I just added a couple of words here, but I not sure if this clarifies anything.

Ergus commented on docs/en/operations/system-tables/codecs.md:
```diff
@@ -0,0 +1,47 @@
+
+- `codec` ([String](../../sql-reference/data-types/string.md)) — Codec name.
+- `method_byte` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — Byte which indicates codec in compressed file.
+- `compression` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is a compression.
```
There are a couple of changes in this fields. Do them improve this anyhow?

rschu1ze commented on docs/en/operations/system-tables/codecs.md:
```diff
@@ -0,0 +1,44 @@
+The `system.codecs` table contains the following columns (the column type is shown in brackets):
+
+- `name` ([String](../../sql-reference/data-types/string.md)) — Codec name.
+- `is_compression` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec performs data compression.
```
Users will not understand the distinction between `is_compression` and `is_generic_compression`. Two options:
- either omit one of the fields, or
- make the description ("The codec performs data compression", ...) more illustrative.

rschu1ze commented on src/Compression/CompressionCodecDelta.cpp:
```diff
@@ -29,6 +29,12 @@ class CompressionCodecDelta : public ICompressionCodec
 
+    std::string getNotes() const override
+    {
+        return "Preprocessor (should be followed by some compression codec). Stores difference between neighboring values; good for monotonically increasing data.";
```
`increasing or decreasing`

rschu1ze commented on src/Compression/ICompressionCodec.h:
```diff
@@ -127,6 +127,8 @@ class ICompressionCodec : private boost::noncopyable
     /// If it does nothing.
     virtual bool isNone() const { return false; }
 
+    virtual std::string getNotes() const = 0;
```
\+ a simple doxygen string like l. 127 etc.

Ergus replied:
Ok

Ergus commented on src/Compression/CompressionCodecEncrypted.h:
```diff
@@ -107,19 +107,13 @@ class CompressionCodecEncrypted final : public ICompressionCodec
+    std::string getNotes() const override
     {
-        return true;
+        return "Encrypts and decrypts blocks with AES-128 in GCM-SIV mode (RFC-8452).";
```
I got these lines liretaly from this: https://kb.altinity.com/altinity-kb-schema-design/codecs/

Ergus committed:
CR2 @rschu1ze

(getNotes -> getDescription): Function renamed
Fix some typos
Re-expose the `method_byte` as suggested by @alexey-milovidov

rschu1ze commented on docs/en/operations/system-tables/codecs.md:
```diff
@@ -0,0 +1,44 @@
+- `is_generic_compression` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is a generic compression.
+- `is_encryption` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec encrypts the data.
+- `is_timeseries_codec`([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is for floating point timeseries codec.
+- `is_delta_compression` ([UInt8](/sql-reference/data-types/int-uint#integer-ranges)) — The codec is for delta compression.
```
Exposing only `is_delta_compression` will not be systematic. Please see the "specialized codecs" section in the [docs](https://clickhouse.com/docs/sql-reference/statements/create/table#specialized-codecs). The question to ask is: what makes the delta codecs special enough to expose a field in the system table but the GCD not - which is another preprocessing codec.
Therefore, my suggestion is to either omit the field or change it to "is_preprocessing_codec". The former is fine as we can always add fields later on but not correct mistakes made during schema design.

rschu1ze commented:
@Ergus A comment about the development process: would it be possible to _not_ force-push after someone (no matter who) reviewed a PR? Force-pushing may (in theory) rewrite the entire PR and from a reviewer perspective it is no longer possible to just check the last added commits (= the diff to the reviewed version), one needs to review the entire PR from scratch.

What is fine is to squash all commits into a single commit at the very end and force-push.

Ergus committed:
CR3 @rschu1ze

Remove `is_delta_compression`

Ergus committed:
CR4 @rschu1ze

* src/Storages/System/StorageSystemCodecs.cpp (getColumnsDescription) : Improve is_compression and is_generic_compression descriptions

Ergus commented:
> Lgtm.
> 
> @Blargian I added an item at [ClickHouse/clickhouse-docs#2730](https://github.com/ClickHouse/clickhouse-docs/issues/2730) so we don't forget to replace the current codec documentation by generated documented based on `system.codecs`.

@Blargian I made the documentation as condensed as possible, so, when you perform the substitution probably is a good idea to extend it to include all the information available currently in docs.

rschu1ze commented:
<img width="1107" alt="grafik" src="https://github.com/user-attachments/assets/10b91043-0c48-4da5-a154-522187a9ac09" />

No, [please no](https://github.com/ClickHouse/ClickHouse/pull/81600#issuecomment-2966659289) 😄

Besides that, lgtm (I formally approved already above).

 Next step: Go through the CI job results and post a small summary report like this [one](https://github.com/ClickHouse/ClickHouse/pull/81752#issuecomment-2977096329) (as an example). Basically: for every failing test, double check if the failure is already registered at github.com/clickhouse/clickhouse/issues, and if no: open a new one. Then, and only then, you may speak the [magic words](https://github.com/rschu1ze/dotfiles/blob/main/.bashrc#L117-L120) to overrule CI.

rschu1ze commented:
> When did I mention to overrule anything?

What I meant is that the CI is not in great shape at the moment (again 😢). Also, as CI only allows to merge PRs with all-green tests, overruling is the only option to get a PR in. To not make CI worse, all test failures need to be reviewed carefully first.

Ergus committed:
Fix test by ordering output

Ergus committed:
Improve last change ordering by ALL

Ergus committed:
Attempt to disable test because ZSTD_QAT and DEFLATE_QPL is x86 only
